File: src/algorithms/trees/construction/build-from-preorder-inorder/sources/build-from-preorder-inorder.py
# Build Binary Tree from Preorder + Inorder Traversal (Recursive)
# First element of preorder is root; find root in inorder to split left/right subtrees

class TreeNode:
    def __init__(self, value: int):
        self.value = value
        self.left = None
        self.right = None

def build_from_preorder_inorder(preorder: list, inorder: list):
    if not preorder or not inorder:  # @step:initialize
        return None

    root_value = preorder[0]  # @step:select-element
    root = TreeNode(root_value)  # @step:build-node

    inorder_root_index = inorder.index(root_value)  # @step:partition-array

    # Split inorder and preorder into left/right sub-arrays
    left_inorder = inorder[:inorder_root_index]  # @step:partition-array
    left_preorder = preorder[1:1 + len(left_inorder)]  # @step:partition-array

    right_inorder = inorder[inorder_root_index + 1:]  # @step:partition-array
    right_preorder = preorder[1 + len(left_inorder):]  # @step:partition-array

    root.left = build_from_preorder_inorder(left_preorder, left_inorder)  # @step:connect-child
    root.right = build_from_preorder_inorder(right_preorder, right_inorder)  # @step:connect-child

    return root  # @step:visit

File: src/algorithms/trees/construction/build-from-preorder-inorder/sources/BuildFromPreorderInorder.java
// Build Binary Tree from Preorder + Inorder Traversal (Recursive)
// First element of preorder is root; find root in inorder to split left/right subtrees

class TreeNode {
    int value;
    TreeNode left, right;
    TreeNode(int value) { this.value = value; }
}

class BuildFromPreorderInorder {
    public TreeNode buildFromPreorderInorder(int[] preorder, int[] inorder) {
        return build(preorder, 0, preorder.length - 1, inorder, 0, inorder.length - 1); // @step:initialize
    }

    private TreeNode build(int[] preorder, int preStart, int preEnd,
                           int[] inorder, int inStart, int inEnd) {
        if (preStart > preEnd || inStart > inEnd) return null; // @step:initialize

        int rootValue = preorder[preStart]; // @step:select-element
        TreeNode root = new TreeNode(rootValue); // @step:build-node

        // Find root position in inorder array
        int inorderRootIndex = inStart; // @step:partition-array
        while (inorderRootIndex <= inEnd && inorder[inorderRootIndex] != rootValue) {
            inorderRootIndex++; // @step:partition-array
        }

        int leftSubtreeSize = inorderRootIndex - inStart; // @step:partition-array

        // Recurse for left and right subtrees
        root.left = build(preorder, preStart + 1, preStart + leftSubtreeSize,
                          inorder, inStart, inorderRootIndex - 1); // @step:connect-child
        root.right = build(preorder, preStart + leftSubtreeSize + 1, preEnd,
                           inorder, inorderRootIndex + 1, inEnd); // @step:connect-child

        return root; // @step:visit
    }
}
